In [1]:
import pandas as pd
import numpy as np
from sklearn import tree
from sklearn import preprocessing
from sklearn.ensemble import RandomForestClassifier

In [2]:
tree_model=tree.DecisionTreeClassifier(max_depth=8)

In [3]:
bank=pd.read_excel("Bank_Personal_Loan_Modelling.xlsx",sheet_name='Data')

In [4]:
bank.head()

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   

   Personal Loan  Securities Account  CD Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

In [9]:
rf_model=RandomForestClassifier(n_estimators=1000,max_features=2,oob_score=True)

In [10]:
bank.columns

Index(['ID', 'Age', 'Experience', 'Income', 'ZIP Code', 'Family', 'CCAvg',
       'Education', 'Mortgage', 'Personal Loan', 'Securities Account',
       'CD Account', 'Online', 'CreditCard'],
      dtype='object')

In [11]:
features=pd.DataFrame([bank["Age"],bank["Experience"],bank["Income"],bank["Family"],bank["CCAvg"],bank["Education"],bank["Mortgage"],bank["Securities Account"],bank["CD Account"],bank["Online"],bank["CreditCard"]]).T

In [12]:
rf_model.fit(X=features,y=bank["Personal Loan"])

RandomForestClassifier(max_features=2, n_estimators=1000, oob_score=True)

In [13]:
print("OOB Accuracy")
print(rf_model.oob_score_);

OOB Accuracy
0.9882


In [14]:
for feature,imp in zip(features,rf_model.feature_importances_):
    print(feature,imp)

Age 0.045549382637802456
Experience 0.044279330999933156
Income 0.34348189314421607
Family 0.09744566215124903
CCAvg 0.18464367878694418
Education 0.16107632504426514
Mortgage 0.04407357218096249
Securities Account 0.005606861038321185
CD Account 0.05520721994593272
Online 0.008491657868096112
CreditCard 0.010144416202277394


In [15]:
bank.isnull().isnull().sum().sum()

0

In [16]:
predictors=pd.DataFrame([bank["Income"],bank["CCAvg"],bank["Education"],bank["Family"]]).T

In [17]:
tree_model.fit(X=predictors,y=bank["Personal Loan"])

DecisionTreeClassifier(max_depth=8)

In [20]:
with open("Bank_Modelling.dot",'w') as f:
    f=tree.export_graphviz(tree_model,feature_names=["Income","CCAvg","Education","Family"],out_file=f)

In [21]:
tree_model.score(X=predictors,y=bank["Personal Loan"])

0.9956

In [22]:
predictors1=pd.DataFrame([bank["Income"],bank["CCAvg"],bank["Education"]]).T

In [23]:
tree_model.fit(X=predictors1,y=bank["Personal Loan"])

DecisionTreeClassifier(max_depth=8)

In [24]:
with open("Bank_ModellingWithoutFamily.dot",'w') as f:
    f=tree.export_graphviz(tree_model,feature_names=["Income","CCAvg","Education"],out_file=f)

In [25]:
tree_model.score(X=predictors1,y=bank["Personal Loan"])

0.9816

# The first Model has 99.56% Accuracy hence we can use that Bank Modeeling Model to decide if someone is eligible for a personal loan or not¶


In [26]:
from sklearn.tree import export_text

In [27]:
feature=["Income","CCAvg","Education"]

In [28]:
r=export_text(tree_model,feature_names=feature)
print(r)

|--- Income <= 113.50
|   |--- CCAvg <= 2.95
|   |   |--- Income <= 106.50
|   |   |   |--- class: 0
|   |   |--- Income >  106.50
|   |   |   |--- Education <= 1.50
|   |   |   |   |--- CCAvg <= 0.35
|   |   |   |   |   |--- CCAvg <= 0.25
|   |   |   |   |   |   |--- Income <= 112.50
|   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |--- Income >  112.50
|   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |--- CCAvg >  0.25
|   |   |   |   |   |   |--- Income <= 111.50
|   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |--- Income >  111.50
|   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |--- CCAvg >  0.35
|   |   |   |   |   |--- Income <= 109.50
|   |   |   |   |   |   |--- CCAvg <= 1.75
|   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |--- CCAvg >  1.75
|   |   |   |   |   |   |   |--- CCAvg <= 1.90
|   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |--- CCAvg >  1.90
|   |   |   |   |   |   